{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import libraries\n",
    "import numpy as np\n",
    "from Bio import SeqIO\n",
    "from Bio.PDB.Polypeptide import d1_to_index\n",
    "from rdkit.Chem import AllChem\n",
    "from __future__ import print_function\n",
    "import pandas as pd\n",
    "import csv\n",
    "from Bio import Entrez\n",
    "import re\n",
    "\n",
    "def download_info (patterm,database=\"Gene\"):\n",
    "    '''download information from relervant database'''\n",
    " \n",
    "    handle = Entrez.esearch(db=database, term=patterm, retmax=500)\n",
    "    record = Entrez.read(handle)\n",
    "    handle.close()\n",
    " \n",
    "    return record\n",
    "\n",
    "def download_seq (id_array):\n",
    "    '''HAVE GeneID to download seqs'''\n",
    " \n",
    "    result_handle = Entrez.efetch(db=\"nucleotide\", rettype=\"fasta\",  id=id_array,retmode=\"text\")\n",
    "    result = result_handle.read()\n",
    " \n",
    "    return result\n",
    "\n",
    "\"\"\" Screen for biosensor sequences \"\"\"\n",
    "# List of detectable chemicals from https://github.com/brsynth/detectable_metabolites\n",
    "file_info = 'detectable_metabolites_seq2.csv'\n",
    "file_output = 'detectable_metabolites_seq.csv'\n",
    "Entrez.email = 'dr413677671@gmail.com'\n",
    "path='F:/'\n",
    "#with open(file_info) as header_input, open(file_output, 'w') as header_output:\n",
    "#    cw = csv.writer(header_output)\n",
    "#    cw.writerow( ('Name', 'Inchi', 'Gene', 'Organism', 'NCBI_Gene', 'Name', 'Description', 'Comment', 'Annotation') )\n",
    "#    cv = csv.DictReader( filter( lambda row: row[0] != '#', header_input ) )\n",
    "cv=pd.read_csv(path+file_info,header=0,sep=',')\n",
    "#    counter=0\n",
    "record=[]\n",
    "\n",
    "# eliminate unuseful columns\n",
    "cv.drop(['Reference', 'TypeOfExperiment','Comments','TypeOfDetection'], axis=1,inplace=True)\n",
    "# Change columns names\n",
    "cv.rename(columns={'DetectedBy':'Sensor'}, inplace = True)\n",
    "cv = cv[pd.notnull(cv['Sensor'])]\n",
    "cv = cv[pd.notnull(cv['InChI'])]\n",
    "\n",
    "cv['info']=None\n",
    "cv['Seq'] = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1290\n",
      "1291\n",
      "1292\n",
      "1293\n",
      "1294\n",
      "1295\n",
      "1296\n",
      "1297\n",
      "1298\n",
      "1299\n",
      "1300\n",
      "1301\n",
      "1302\n",
      "1303\n",
      "1304\n",
      "1305\n",
      "1306\n",
      "1307\n",
      "1308\n",
      "1309\n",
      "1310\n",
      "1311\n",
      "1312\n",
      "1313\n",
      "1314\n",
      "1315\n",
      "1316\n",
      "1317\n",
      "1318\n",
      "1319\n",
      "1320\n",
      "1321\n",
      "1322\n",
      "1323\n",
      "1324\n",
      "1325\n",
      "1326\n",
      "1327\n",
      "1328\n",
      "1329\n",
      "1330\n",
      "1331\n",
      "1332\n",
      "1333\n",
      "1334\n",
      "1335\n",
      "1336\n",
      "1337\n",
      "1338\n",
      "1339\n",
      "1340\n",
      "1341\n",
      "1342\n",
      "1343\n",
      "1344\n",
      "1345\n",
      "1346\n",
      "1347\n",
      "1348\n",
      "1349\n",
      "1350\n",
      "1351\n",
      "1352\n",
      "1353\n",
      "1354\n",
      "1355\n",
      "1356\n",
      "1357\n",
      "1358\n",
      "1359\n",
      "1360\n",
      "1361\n",
      "1362\n",
      "1363\n",
      "1364\n",
      "1365\n",
      "1366\n",
      "1367\n",
      "1368\n",
      "1369\n",
      "1370\n",
      "1371\n",
      "1372\n",
      "1373\n",
      "1374\n",
      "1375\n",
      "1376\n",
      "1377\n",
      "1378\n",
      "1379\n",
      "1380\n",
      "1381\n",
      "1382\n",
      "1383\n",
      "1384\n",
      "1385\n",
      "1386\n",
      "1387\n",
      "1388\n",
      "1389\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "if not os.path.exists(path):\n",
    "        os.mkdir(path)\n",
    "#5*278\n",
    "path='F:/'\n",
    "for i in range(1290,1390,1):\n",
    "#     print((i+1)*109)\n",
    "    tmp=cv[i:(i+1)]\n",
    "   \n",
    "    for index, row in tmp.iterrows():\n",
    "        print(i)\n",
    "        name = str(row['Name'])\n",
    "        inchi = row['InChI']\n",
    "        organism = str(row['Organism'])\n",
    "        sensor = str(row['Sensor'])\n",
    "        # Query based on gene name\n",
    "        # TO DO: curate generic names (riboswitch, TF, etc.)\n",
    "        if len(sensor) > 0:\n",
    "            # Query NCBI by gene name\n",
    "            term=\"\"\n",
    "            term=term+ sensor+' [GN]' \n",
    "            if organism!= 'nan':\n",
    "                term=term+ ' AND ' + organism+' [ORGN]' \n",
    "            try:\n",
    "                record=download_info(term,database=\"Nucleotide\") \n",
    "            except:\n",
    "                continue\n",
    "            # Fetch the full record or just the summary (should be enough)\n",
    "            # Things to do:\n",
    "            # - Select the most convenient format (xml, etc.)\n",
    "            # - Query by name to avoid false hits\n",
    "            # - Use Description to double check transcriptional activity\n",
    "            # - Double-check organims (better through postprocessing)\n",
    "            if  record['IdList']:\n",
    "                id_array =download_seq( record['IdList'][0] )\n",
    "#                 id_info.append(id_array)\n",
    "                try:\n",
    "                    infotmp,seqtmp=np.squeeze(re.findall(r'>(.*)\\n([\\s\\S]*$)',id_array))\n",
    "                    cv.loc[index:index, 'info']=infotmp.encode('utf-8').strip()\n",
    "                    cv.loc[index:index, 'Seq']=re.sub(\"\\n\",\"\",seqtmp)\n",
    "                except:\n",
    "                    continue\n",
    "            \n",
    "            \n",
    "        # add genes info as new columns to dataframe\n",
    "\n",
    "# write csv\n",
    "\n",
    "file_output = path + 'detectable_metabolites_seq2.csv'\n",
    "cv.to_csv(file_output,index=True,sep=',')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "file_output = path + 'detectable_metabolites_seq2.csv'\n",
    "cv.to_csv(file_output,index=True,sep=',')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
